### PR TITLE
Typecheck refactor for `!`

### PR DESCRIPTION
**Pull request by canndrew**
Ping @nikomatsakis @eddyb. This is the PR for the typeck refactor for `!`. Is this what you guys had in mind? Is there anything else that needs doing on it?

### CODE REVIEW

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -38,12 +38,11 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 self.write_ty(pat.id, expected);
             }
             PatKind::Lit(ref lt) => {
-                self.check_expr(&lt);
-                let expr_ty = self.expr_ty(&lt);
+                let expr_t = self.check_expr(&lt);
```
Comment (eddyb) [inline src/librustc_typeck/check/_match.rs]: Why this change? `expr_ty` is a better name for such a variable.

File: src/librustc_typeck/check/mod.rs [hunk 2]
```diff
@@ -3268,22 +3239,22 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
         };
 
-        self.check_expr_struct_fields(expr_ty, path.span, variant, fields,
+        self.check_expr_struct_fields(expr_t, path.span, variant, fields,
                                       base_expr.is_none());
         if let &Some(ref base_expr) = base_expr {
-            self.check_expr_has_type(base_expr, expr_ty);
-            match expr_ty.sty {
+            self.check_expr_has_type(base_expr, expr_t);
+            match expr_t.sty {
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Same `expr_t` here - is this an artifact from automated search/replace?

File: src/librustc_typeck/check/_match.rs [hunk 3]
```diff
@@ -605,6 +601,7 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
 
         // Type check the path.
         let pat_ty = self.instantiate_value_path(segments, opt_ty, def, pat.span, pat.id);
+        self.write_ty(pat.id, pat_ty);
```
Comment (petrochenkov) [inline src/librustc_typeck/check/_match.rs]: Why pulling `write_ty` out of things?
(I moved this common code into `instantiate_value_path`/`instantiate_type_path` just recently.)

**Comment by arielb1** [comment]
I would move the match in `check_expr_with_expectation_and_lvalue_pref` to its own function, to prevent confusion if someone introduces a `return` there.

File: src/librustc_typeck/check/mod.rs [hunk 4]
```diff
@@ -3507,66 +3478,54 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+            self.check_match(expr, &discrim, arms, expected, match_src)
           }
           hir::ExprClosure(capture, ref decl, ref body, _) => {
-              self.check_expr_closure(expr, capture, &decl, &body, expected);
+              self.check_expr_closure(expr, capture, &decl, &body, expected)
           }
           hir::ExprBlock(ref b) => {
             self.check_block_with_expected(&b, expected);
-            self.write_ty(id, self.node_ty(b.id));
+            self.node_ty(b.id)
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: `check_block_with_expected` should return the type.

File: src/librustc_typeck/check/mod.rs [hunk 5]
```diff
@@ -3926,17 +3882,15 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
           hir::StmtExpr(ref expr, id) => {
             node_id = id;
             // Check with expected type of ()
-            self.check_expr_has_type(&expr, self.tcx.mk_nil());
-            let expr_ty = self.expr_ty(&expr);
-            saw_bot = saw_bot || self.type_var_diverges(expr_ty);
-            saw_err = saw_err || expr_ty.references_error();
+            let expr_t = self.check_expr_has_type(&expr, self.tcx.mk_nil());
+            saw_bot = saw_bot || self.type_var_diverges(expr_t);
+            saw_err = saw_err || expr_t.references_error();
```
Comment (eddyb) [inline src/librustc_typeck/check/mod.rs]: Same `expr_t` name change.

File: src/librustc_typeck/check/_match.rs (hunk 1 above)
Comment (canndrew) [inline src/librustc_typeck/check/_match.rs]: Because having a method named `expr_ty` and a variable named `expr_ty` is confusing.

Comment (eddyb) [inline src/librustc_typeck/check/_match.rs]: I find `_t` worse. Besides, why do we still need the `expr_ty` method?

File: src/librustc_typeck/check/_match.rs (hunk 3 above)
Comment (canndrew) [inline src/librustc_typeck/check/_match.rs]: I wanted to pull all the `write_ty`s down to one spot so that types for expressions always get written in the same place. Then I could add the adjustment in that spot.

File: src/librustc_typeck/check/_match.rs (hunk 1 above)
Comment (canndrew) [inline src/librustc_typeck/check/_match.rs]: `expr_ty` is still used in a few places. I can keep going trying to remove it entirely if you like. Otherwise I can rename the variable `ty`.

Comment (eddyb) [inline src/librustc_typeck/check/_match.rs]: Either sound good, thanks!

File: src/librustc_typeck/check/_match.rs (hunk 3 above)
Comment (eddyb) [inline src/librustc_typeck/check/_match.rs]: Maybe patterns should have something similar?

**Comment by canndrew** [comment]
@eddyb is this PR good to go? I can do the same thing with patterns (factoring out `write_ty`) if you'd like.

**Comment by eddyb** [comment]
@canndrew Yeah, doing that would be great. Are there many `expr_ty` calls left, btw?

**Comment by canndrew** [comment]
I've done the factoring-out for patterns. There's still some uses of `expr_ty` left though most of them look like they'll be non-trivial to get rid of (the ones in coercion.rs and confirm.rs).

**Comment by eddyb** [comment]
@canndrew Can't both of those takes the types in? Although not confirm.rs if it's a late adjustment.
That is, I'd like to remove uses of `expr_ty` that are done *near* the expression being type-checked, and keep only those that are deferred, possibly manually getting the node type if there's only 2-3 of them.

**Comment by canndrew** [comment]
I've made the coercions code take the types as arguments.

The code in confirm.rs walks over a big expression tree reading the types using `expr_ty`. So there's no way to pass the types in there. I could replace the `expr_ty`s with `node_ty`s but it means that code will no longer be able to see the `AdjustNeverToAny` coercion (does that matter?).

File: src/librustc_typeck/check/_match.rs [hunk 6]
```diff
@@ -414,11 +414,14 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             }
             _ => result_ty
         };
+
+        let mut arm_tys = Vec::new();
```
Comment (eddyb) [inline src/librustc_typeck/check/_match.rs]: This shouldn't be needed though. The design was such that no extra bookkeeping was needed.

File: src/librustc_typeck/check/coercion.rs [hunk 7]
```diff
@@ -735,13 +737,13 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
         // previous expressions, other than noop reborrows (ignoring lifetimes).
-        for expr in exprs() {
+        for (expr, expr_ty) in exprs() {
             let noop = match self.tables.borrow().adjustments.get(&expr.id) {
                 Some(&AdjustDerefRef(AutoDerefRef {
                     autoderefs: 1,
                     autoref: Some(AutoPtr(_, mutbl_adj)),
                     unsize: None
-                })) => match self.expr_ty(expr).sty {
+                })) => match expr_ty.sty {
```
Comment (eddyb) [inline src/librustc_typeck/check/coercion.rs]: This should actually read the type from the map holding those types, just like it reads the adjustment.

Comment (canndrew) [inline src/librustc_typeck/check/coercion.rs]: Ah I see, I don't need to thread the types through then.

**Comment by canndrew** [comment]
Any idea what these LLVM linker errors are about?

**Comment by canndrew** [comment]
@eddyb I rebased this and now it's passing. Is it good to merge?
